src/MinimalKafka/KafkaProducerFactory.cs

```diff
@@ -1,34 +1,83 @@
 ﻿using Confluent.Kafka;
+using Confluent.Kafka.Admin;
 using Microsoft.Extensions.DependencyInjection;
 using MinimalKafka.Builders;
 using MinimalKafka.Metadata;
+using System.Globalization;
+using TopicMetadata = MinimalKafka.Metadata.TopicMetadata;
 
 namespace MinimalKafka;
 
-public class KafkaProducerFactory<TKey, TValue> : IProducer<TKey, TValue>
+public class KafkaProducerFactory<TKey, TValue> : IProducer<TKey, TValue>, IMinimalProducer<TKey, TValue>
 {
+    private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(5);
+    
+    private readonly IAdminClient _adminClient;
+    
+    private readonly ITopicMetadata _topicMetadata;
+
+    private bool _initialized;
+
     public IProducer<TKey, TValue> Producer { get; set; }
 
     public Handle Handle => Producer.Handle;
 
     public string Name => Producer.Name;
 
-    public KafkaProducerFactory(IKafkaBuilder builder)
+    public KafkaProducerFactory(IKafkaBuilder builder, IAdminClient adminClient)
     {
+        _adminClient = adminClient;
         var config = builder.MetaData.OfType<ConfigurationMetadata>().First();
         var keySerializer = builder.MetaData.OfType<KeySerializerMetadata>().First(); 
         var valueSerializer = builder.MetaData.OfType<ValueSerializerMetadata>().First();
         var producerConfig = new ProducerConfig(config.Configuration);
 
         var serializerKey = ActivatorUtilities.CreateInstance(builder.ServiceProvider, keySerializer.GetSerializerType<TKey>());
         var serializerValue = ActivatorUtilities.CreateInstance(builder.ServiceProvider, valueSerializer.GetSerializerType<TValue>());
-
+        
+        _topicMetadata = builder.MetaData.OfType<TopicMetadata<TValue>>().FirstOrDefault()
+                ?? builder.MetaData.OfType<TopicMetadata>().First();
+        
         Producer = new ProducerBuilder<TKey, TValue>(producerConfig)
             .SetKeySerializer((ISerializer<TKey>)serializerKey)
             .SetValueSerializer((ISerializer<TValue>)serializerValue)
             .Build();
     }
 
+    private async Task Initialize()
+    {
+        if (_initialized) return;
```

Comment:
If you agree on the general concept I can spend more time refactoring this into it's own class.

Reply:
Ideally this is done before the application is started, aka `IHostedService.StartAsync`.

Reply:
I fully agree with the idea but we should treat i like a first class citizen and we can Run it in `KafkaService.StartAsync` before the KafkaProcesses are started